### PR TITLE
OESA Foundation - Create Application Context

### PR DESCRIPTION
**Pull request by jorg3lopez**
This PR adds the Application Context class that behaves like a factory and it will be used for injecting implementations via annotations. 

#### Changes
- Added ApplicationContext.java class
- Added ApplicationContext unit test (ApplicationContextTest)

### CODE REVIEW

File: app/src/test/groovy/gov/hhs/cdc/trustedintermediary/ApplicationContextTest.groovy [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+        setup:
+        def contextA = ApplicationContext.getInstance()
+        def contextB = ApplicationContext.getInstance()
+
+        when:
+        def result = contextA == contextB?true:false
+
+        then:
+        result == true
+    }
```
Comment (jorg3lopez) [inline app/src/test/groovy/gov/hhs/cdc/trustedintermediary/ApplicationContextTest.groovy:16]: I want to add a test case to test the retrieval of the implementations. Currently, we have no classes to test this with, but I'm open to suggestions. Let me know your thoughts.

Comment (halprin) [reply app/src/test/groovy/gov/hhs/cdc/trustedintermediary/ApplicationContextTest.groovy:16]: You could test it with a standard Java class. For example,

```java
applicationContext.register(String.class, "DogCow");
```

**Comment by halprin** [comment]
I can tell while looking at this PR's description on my mobile phone whether #13 will be closed upon merge, but this PR will not fulfill all the acceptance criteria of #13. So, I'd say remove the "resolves" thing.

File: app/src/main/java/gov/hhs/cdc/trustedintermediary/context/ApplicationContext.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    public static <T> T getImplementation(Class<T> clazz) {
+        return (T) OBJECT_MAP.get(clazz);
+    }
+
+    public static ApplicationContext getInstance() {
+        if (applicationContext == null) {
+            synchronized (ApplicationContext.class) {
+                if (applicationContext == null) applicationContext = new ApplicationContext();
```
Comment (halprin) [inline app/src/main/java/gov/hhs/cdc/trustedintermediary/context/ApplicationContext.java:30]: Mind putting curly braces on this if statement? I think we should always use them. I'm actually a little surprised that the code formatter allowed this.

File: app/src/main/java/gov/hhs/cdc/trustedintermediary/context/ApplicationContext.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
+/**
+ * ApplicationContext class works similar to a factory. The idea is to use this class in conjunction
+ * with annoations that will be used to inject the implementations. There is the option of using the
```
Comment (halprin) [inline app/src/main/java/gov/hhs/cdc/trustedintermediary/context/ApplicationContext.java:3]: Just a small typo.
```suggestion
 * with annotations that will be used to inject the implementations. There is the option of using the
```

**Review by halprin — COMMENTED** [review]
Nice threading protection. I just have a few comments, but this looks great!

File: app/src/test/groovy/gov/hhs/cdc/trustedintermediary/ApplicationContextTest.groovy (hunk 1 above)
Comment (jorg3lopez) [reply app/src/test/groovy/gov/hhs/cdc/trustedintermediary/ApplicationContextTest.groovy:16]: Yes! We think alike. I was thinking in those parameters. I'll update it when I'm able to, thanks!

File: app/src/main/java/gov/hhs/cdc/trustedintermediary/context/ApplicationContext.java (hunk 2 above)
Comment (jorg3lopez) [reply app/src/main/java/gov/hhs/cdc/trustedintermediary/context/ApplicationContext.java:30]: I'll get on it. I think I had it with curly braces, but the formatter changed it. I'll double check.

Comment (jorg3lopez) [reply app/src/main/java/gov/hhs/cdc/trustedintermediary/context/ApplicationContext.java]: I was able to add the curly braces!

File: app/src/main/java/gov/hhs/cdc/trustedintermediary/context/ApplicationContext.java (hunk 3 above)
Comment (jorg3lopez) [reply app/src/main/java/gov/hhs/cdc/trustedintermediary/context/ApplicationContext.java]: Good catch!

File: app/src/test/groovy/gov/hhs/cdc/trustedintermediary/ApplicationContextTest.groovy [hunk 4]
```diff
@@ -0,0 +1,27 @@
+package gov.hhs.cdc.trustedintermediary.context
```
Comment (halprin) [inline app/src/test/groovy/gov/hhs/cdc/trustedintermediary/ApplicationContextTest.groovy:1]: I just realized, this file is in the wrong folder.  It also needs to be added into a `context` sub-folder.

Comment (jorg3lopez) [reply app/src/test/groovy/gov/hhs/cdc/trustedintermediary/ApplicationContextTest.groovy:1]: Good catch, I imported the class.

Comment (halprin) [reply app/src/test/groovy/gov/hhs/cdc/trustedintermediary/ApplicationContextTest.groovy]: I meant you should go the other way.  Keep the `.context` on the `package` line and move this file into a newly created `context` sub-folder so it mirrors the `src/main`.

**Pull request merged**